### PR TITLE
Fuzzy table matching part 2

### PR DESCRIPTION
**Pull request by ahuang11**

### CODE REVIEW

File: lumen/ai/assistant.py [hunk 1]
```diff
@@ -213,6 +214,8 @@ async def _invalidate_memory(self, messages):
             invalid_key = validity.invalid_key
             if invalid_key == "table":
                 memory.pop("current_table", None)
+                memory.pop("current_data", None)
```
Comment (ahuang11) [inline lumen/ai/assistant.py:217]: Pop out stale data

File: lumen/ai/assistant.py [hunk 2]
```diff
@@ -182,8 +182,9 @@ def use_suggestion(event):
     async def _invalidate_memory(self, messages):
         table = memory.get("current_table")
         sql = memory.get("current_sql")
-        if not sql:
+        if not table and not sql:
```
Comment (ahuang11) [inline lumen/ai/assistant.py:185]: Be sure to invalidate table too.

File: lumen/ai/agents.py [hunk 3]
```diff
@@ -332,17 +343,6 @@ async def requirements(self, messages: list | str):
...
-        tables = source.get_tables() if source else []
-        if len(tables) > FUZZY_TABLE_LENGTH:
-            closest_tables = await self._get_closest_tables(messages, tables, n=5)
-            if len(closest_tables) == 0:
-                # if no tables are found, ask the user to select ones and load it
-                tables = await self._select_table(tables)
-                return self.requires + ['current_table']
-            elif len(closest_tables) == 1:
-                return self.requires + ['current_table']
-
```
Comment (ahuang11) [inline lumen/ai/agents.py:345]: Not needed with better logic in invalidating and checks.

File: lumen/ai/agents.py [hunk 4]
```diff
@@ -196,17 +196,28 @@ async def _get_closest_tables(self, messages: list | str, tables: list[str], n:
         for keyword in table_keywords:
             closest_tables |= set(difflib.get_close_matches(keyword, list(tables_lower), n=5, cutoff=0.3))
         closest_tables = [tables_lower[table] for table in closest_tables]
-        return tuple(closest_tables)
+        if len(closest_tables) == 0:
+            # if no tables are found, ask the user to select ones and load it
+            tables = await self._select_table(tables)
```
Comment (ahuang11) [inline lumen/ai/agents.py:201]: Only select if unavailable to auto select.

**Review by philippjfr — APPROVED** [review]

**Pull request merged**